crates/project/src/lsp_store.rs

```diff
@@ -5812,6 +5823,30 @@ impl LspStore {
         };
 
         if let Some(buffer) = self.buffer_store.read(cx).get_by_path(&project_path, cx) {
+            let snapshot = self
+                .as_local_mut()
+                .unwrap()
+                .buffer_snapshot_for_lsp_version(&buffer, server_id, version, cx)?;
+
+            if let Some(existing_diagnostics) =
+                buffer.read(cx).get_diagnostics(server_id).map(|diag| {
+                    diag.iter()
+                        .filter(|v| filter(&v.diagnostic))
+                        .map(|v| {
+                            let start = Unclipped(v.range.start.to_point_utf16(&snapshot));
+                            let end = Unclipped(v.range.end.to_point_utf16(&snapshot));
+                            let range = start..end;
+                            DiagnosticEntry {
+                                range,
+                                diagnostic: v.diagnostic.clone(),
+                            }
+                        })
+                        .collect::<Vec<_>>()
```

Comment:
```suggestion
```

Can also do `.get_diagnostics(server_id).into_iter().flat_map(...` and avoid any extra nesting and `if let Some`.